### PR TITLE
Twl/range queries and delete

### PR DESCRIPTION
**Pull request by Tadeuchi**

### CODE REVIEW

File: src/LmdbCache.ts [hunk 1]
```diff
@@ -216,4 +254,63 @@ export class LmdbCache<V = any> implements SortKeyCache<V> {
...
+
+    return result;
+  }
+
+  async rollback(): Promise<void> {
+    await this.db.transactionSync(this.rollbackBatch);
+  }
+}
+
+class ClientValueWrapper<V> {
```
Comment (ppedziwiatr) [inline src/LmdbCache.ts:314]: shouldn't this be imported from the SDK?

File: src/LmdbCache.ts [hunk 2]
```diff
@@ -30,113 +34,153 @@ export class LmdbCache<V = any> implements SortKeyCache<V> {
...
-        .forEach((key) => {
-          this.db.remove(key);
-        });
-    });
+  private async doPut(cacheKey: CacheKey, value: ClientValueWrapper<V>): Promise<boolean> {
+    const putResult = await this.db.put(this.dbEntryKey(cacheKey), value);
+    if (putResult) {
+      const previousCalls = this.rollbackBatch;
+      const db = this.db;
+      this.rollbackBatch = () => {
```
Comment (ppedziwiatr) [inline src/LmdbCache.ts:123]: I'm not sure I like this solution...this.rollbackBatch is class field (for an instance that can be considered a 'signleton' in a warp instance context), isn't there a risk that during contract evaluation (if we're reading/writing states to/from other contracts) - we will overwrite this value before the previous one was effectively used?

File: src/LmdbCache.ts [hunk 3]
```diff
@@ -30,113 +34,153 @@ export class LmdbCache<V = any> implements SortKeyCache<V> {
...
+   * Batch operations are all executed in one transaction using childTransaction.
+   * For each put there is an old entries removal run.
+   */
   async batch(opStack: BatchDBOp<V>[]) {
-    for (const op of opStack) {
-      if (op.type === 'put') {
-        await this.put(op.key, op.value);
-      } else if (op.type === 'del') {
-        await this.delete(op.key);
+    await this.db.childTransaction(async () => {
```
Comment (ppedziwiatr) [inline src/LmdbCache.ts:48]: not sure if async code within a transactions is safe
image

File: src/LmdbCache.ts (hunk 2 above)
Comment (ppedziwiatr) [reply src/LmdbCache.ts:123]: re. 'singleton' - I'm talking about the json-based state - where all contracts' state is stored in a single db file and instance.

File: src/LmdbCache.ts [hunk 4]
```diff
@@ -158,14 +202,8 @@ export class LmdbCache<V = any> implements SortKeyCache<V> {
...
-  async keys(): Promise<string[]> {
-    const keys = this.db.getKeys();
-    const contracts = new Set<string>();
-    keys.forEach((k) => {
-      contracts.add(k.split('|')[0]);
-    });
-
-    return Array.from(contracts);
+  async keys(sortKey?: string, options?: SortKeyCacheRangeOptions): Promise<string[]> {
+    return Array.from((await this.kvMap(sortKey, options)).keys());
```
Comment (erhant) [inline src/LmdbCache.ts:206]: The type of `sortKey` in `kvMap` is `sortKey: string`, but we are providing the type `sortKey?: string` to that argument in this call. Is this safe?

File: src/LmdbCache.ts (hunk 1 above)
Comment (Tadeuchi) [reply src/LmdbCache.ts:314]: Well..., maybe, i dont know...
ClientValueWrapper is a helper class to implement delete functionality in leveldb. 
It is very much related to how we use leveldb, but I don't know if we should export it as part of the api.

It so happens that lmdb also implements delete in the similar way, so this class looks the same, but it doesn't have to by like that. 
Maybe the ClientValueWrapper for leveldb will change in a different way then the one for the lmdb.

File: src/LmdbCache.ts (hunk 3 above)
Comment (Tadeuchi) [reply src/LmdbCache.ts:48]: yes, it seems that using transaction or childTransaction functions we may end up with some unpredicted order...
maybe we should switch to transactionSync

File: src/LmdbCache.ts (hunk 4 above)
Comment (Tadeuchi) [reply src/LmdbCache.ts:206]: I think I missed something here, I'll look into it :/

File: src/LmdbCache.ts (hunk 1 above)
Comment (ppedziwiatr) [reply src/LmdbCache.ts:321]: Ok.

**Review by ppedziwiatr — APPROVED** [review]

**Pull request merged**